File: BE/models/Voucher.js
const mongoose = require('mongoose');

const voucherSchema = new mongoose.Schema({
  voucherId: {
    type: String,
    unique: true,
    required: true
  },
  code: {
    type: String,
    unique: true,
    required: true,
    uppercase: true,
    trim: true
  },
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  discountValue: {
    type: Number,
    required: true,
    min: 0
  },
  discountType: {
    type: String,
    enum: ['percentage', 'fixed'],
    required: true
  },
  minOrderValue: {
    type: Number,
    default: 0
  },
  maxDiscountAmount: {
    type: Number
  },
  expiryDate: {
    type: Date,
    required: true
  },
  isActive: {
    type: Boolean,
    default: true
  },
  applicableCategory: {
    type: String,
    enum: ['hotel', 'tour', 'flight', 'restaurant', 'all'],
    default: 'all'
  },
  usageLimit: {
    type: Number,
    default: null
  },
  usedCount: {
    type: Number,
    default: 0
  }
}, {
  timestamps: true
});

voucherSchema.index({ code: 1 });
voucherSchema.index({ expiryDate: 1, isActive: 1 });

voucherSchema.methods.isValid = function() {
  const now = new Date();
  return this.isActive && 
         this.expiryDate > now && 
         (this.usageLimit === null || this.usedCount < this.usageLimit);
};

voucherSchema.methods.calculateDiscount = function(orderValue) {
  if (!this.isValid()) return 0;
  if (orderValue < this.minOrderValue) return 0;

  let discount = 0;
  if (this.discountType === 'percentage') {
    discount = (orderValue * this.discountValue) / 100;
    if (this.maxDiscountAmount && discount > this.maxDiscountAmount) {
      discount = this.maxDiscountAmount;
    }
  } else {
    discount = this.discountValue;
  }
  return discount;
};

module.exports = mongoose.model('Voucher', voucherSchema);

File: fe/lib/models/voucher.dart
import 'package:flutter/material.dart';

class Voucher {
  final String id;
  final String code;
  final String title;
  final String description;
  final double discountValue;
  final String discountType; // 'percentage' or 'fixed'
  final double? minOrderValue;
  final DateTime expiryDate;
  final bool isActive;
  final String? applicableCategory;

  Voucher({
    required this.id,
    required this.code,
    required this.title,
    required this.description,
    required this.discountValue,
    required this.discountType,
    required this.expiryDate,
    this.minOrderValue,
    this.isActive = true,
    this.applicableCategory,
  });

  factory Voucher.fromJson(Map<String, dynamic> json) {
    return Voucher(
      id: json['_id'] ?? '',
      code: json['code'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      discountValue: (json['discountValue'] ?? 0).toDouble(),
      discountType: json['discountType'] ?? 'percentage',
      expiryDate: DateTime.parse(
        json['expiryDate'] ?? DateTime.now().toIso8601String(),
      ),
      minOrderValue: json['minOrderValue']?.toDouble(),
      isActive: json['isActive'] ?? true,
      applicableCategory: json['applicableCategory'],
    );
  }

  Color get color {
    switch (applicableCategory?.toLowerCase()) {
      case 'hotel':
        return Colors.blue;
      case 'tour':
        return Colors.green;
      case 'flight':
        return Colors.orange;
      default:
        return Colors.purple;
    }
  }

  bool get isExpired => DateTime.now().isAfter(expiryDate);

  String get discountText {
    if (discountType == 'percentage') {
      return '${discountValue.toInt()}%';
    } else {
      return '${discountValue.toStringAsFixed(0)}đ';
    }
  }
}
